Mathlib/LinearAlgebra/Matrix/SVD/Defs.lean

```diff
@@ -0,0 +1,479 @@
+    not_false_eq_true, diagonal_apply_ne]
+  simp_rw [← Matrix.ext_iff, of_apply, zero_apply, implies_true, true_and ]
+  intro i j
+  by_cases i = j
+  · unfold eigenColumnEquiv Equiv.sumCongr
+    simp only [ne_eq, Equiv.symm_trans_apply, Equiv.symm_symm, Equiv.coe_fn_symm_mk, Sum.elim_inr,
+      Equiv.sumCompl_apply_inr]
+    simp_rw [h, diagonal_apply_eq]
+    apply eigen_eigenColumnEquiv_inr
+  · simp only [diagonal_apply_ne, ne_eq, EmbeddingLike.apply_eq_iff_eq, Sum.inr.injEq, h]
+
+lemma S'_zero_blocks (A : Matrix (Fin M) (Fin N) 𝕂) :
+    A.svdS'.toBlocks₁₂ = 0 ∧ A.svdS'.toBlocks₂₁ = 0 ∧ A.svdS'.toBlocks₂₂ = 0 := by
+  unfold toBlocks₁₂ toBlocks₂₁ toBlocks₂₂ svdS'
+  simp only [reindex_apply, submatrix_apply, ne_eq, EmbeddingLike.apply_eq_iff_eq,
+    not_false_eq_true, diagonal_apply_ne]
+  simp_rw [← Matrix.ext_iff, of_apply, zero_apply, implies_true, true_and ]
+  intro i j
+  by_cases i = j
+  · unfold eigenRowEquiv Equiv.sumCongr
+    simp only [ne_eq, Equiv.symm_trans_apply, Equiv.symm_symm, Equiv.coe_fn_symm_mk, Sum.elim_inr,
+      Equiv.sumCompl_apply_inr]
+    simp_rw [h, diagonal_apply_eq]
+    apply eigen_eigenRowEquiv_inr
+  · simp only [diagonal_apply_ne, ne_eq, EmbeddingLike.apply_eq_iff_eq, Sum.inr.injEq, h]
+
+lemma S_σpos_block (A : Matrix (Fin M) (Fin N) 𝕂) :
+    A.svdS.toBlocks₁₁ = A.svdμ := by
+  unfold toBlocks₁₁ svdμ svdS
+  simp only [reindex_apply, submatrix_apply, ne_eq, EmbeddingLike.apply_eq_iff_eq, Sum.inl.injEq,
+    submatrix_diagonal_equiv]
+  funext i j
+  by_cases h: i=j
+  · simp_rw [h]
+    unfold eigenColumnEquiv finRankEquivEigsConjTransposeMulSelf
+      Equiv.sumCongr Sum.map
+    simp only [ne_eq, Equiv.symm_trans_apply, Equiv.symm_symm, Equiv.coe_fn_symm_mk,
+      Sum.elim_inl, Equiv.sumCompl_apply_inl, of_apply,
+      diagonal_apply_eq, Function.comp_apply]
+  · rw [diagonal_apply_ne, of_apply, diagonal_apply_ne]
+    rw [ne_eq, EmbeddingLike.apply_eq_iff_eq, Sum.inl.injEq]
+    assumption'
+
+lemma S'_σpos_block (A : Matrix (Fin M) (Fin N) 𝕂) :
+    A.svdS'.toBlocks₁₁ = A.svdμ' := by
+  unfold toBlocks₁₁ svdμ' svdS'
+  simp only [reindex_apply, submatrix_apply, ne_eq, EmbeddingLike.apply_eq_iff_eq, Sum.inl.injEq,
+    submatrix_diagonal_equiv]
+  funext i j
+  by_cases h: i=j
+  · simp_rw [h]
+    unfold eigenRowEquiv finRankEquivEigsMulConjTranspose
+      Equiv.sumCongr Sum.map
+    simp only [ne_eq, Equiv.symm_trans_apply, Equiv.symm_symm, Equiv.coe_fn_symm_mk,
+      Sum.elim_inl, Equiv.sumCompl_apply_inl, of_apply,
+      diagonal_apply_eq, Function.comp_apply]
+  · rw [diagonal_apply_ne, of_apply, diagonal_apply_ne]
+    rw [ne_eq, EmbeddingLike.apply_eq_iff_eq, Sum.inl.injEq]
+    assumption'
+
+lemma S_block (A : Matrix (Fin M) (Fin N) 𝕂) :
+    (reindex
+      (eigenColumnEquiv A)
+      (eigenColumnEquiv A))
+        ( diagonal ( (isHermitian_transpose_mul_self A).eigenvalues)) =
+          fromBlocks A.svdμ 0 0 0 := by
```

Comment:
Isn't the left hand side exactly `A.svdS`? Then I would restate this (and the same for `S'`) as:
```suggestion
lemma fromBlocks_μ_zero_zero_zero (A : Matrix (Fin M) (Fin N) 𝕂) :
  fromBlocks A.svdμ 0 0 0 = A.svdS := by
```

Reply:
I got rid of S and now have `μ_block` and `μ'_block`